File: salesflow-app/src/api/salesBrain.ts
/**
 * 🧠 Sales Brain API Client
 * Teach-UI & Rule Learning System
 */

import { API_CONFIG } from "../services/apiConfig";
import type {
  CreateRulePayload,
  CreateRuleResponse,
  SalesBrainRule,
  GetRulesResponse,
  RuleScope,
} from "./types/salesBrain";

const API_BASE = `${API_CONFIG.baseUrl}/sales-brain`;

// =============================================================================
// RULES
// =============================================================================

/**
 * Erstellt eine neue Sales Brain Regel
 */
export async function createSalesBrainRule(
  payload: CreateRulePayload
): Promise<CreateRuleResponse> {
  const response = await fetch(`${API_BASE}/rules`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    credentials: "include",
    body: JSON.stringify({
      scope: payload.scope,
      override: {
        suggestion_id: payload.override.suggestionId,
        original_text: payload.override.originalText,
        final_text: payload.override.finalText,
        similarity_score: payload.override.similarityScore,
        override_type: payload.override.overrideType,
        context: {
          vertical_id: payload.override.context.verticalId,
          company_id: payload.override.context.companyId,
          channel: payload.override.context.channel,
          use_case: payload.override.context.useCase,
          language: payload.override.context.language,
          lead_status: payload.override.context.leadStatus,
          deal_state: payload.override.context.dealState,
          lead_sentiment: payload.override.context.leadSentiment,
        },
      },
      note: payload.note,
      auto_tag: payload.autoTag,
    }),
  });

  if (!response.ok) {
    const error = await response.json().catch(() => ({}));
    throw new Error(error.detail || `HTTP ${response.status}`);
  }

  return response.json();
}

/**
 * Holt alle Regeln für den aktuellen User/Team
 */
export async function getSalesBrainRules(params?: {
  scope?: RuleScope;
  channel?: string;
  useCase?: string;
  page?: number;
  pageSize?: number;
}): Promise<GetRulesResponse> {
  const searchParams = new URLSearchParams();
  if (params?.scope) searchParams.set("scope", params.scope);
  if (params?.channel) searchParams.set("channel", params.channel);
  if (params?.useCase) searchParams.set("use_case", params.useCase);
  if (params?.page) searchParams.set("page", params.page.toString());
  if (params?.pageSize) searchParams.set("page_size", params.pageSize.toString());

  const url = `${API_BASE}/rules${searchParams.toString() ? `?${searchParams}` : ""}`;

  const response = await fetch(url, {
    credentials: "include",
  });

  if (!response.ok) {
    throw new Error(`HTTP ${response.status}`);
  }

  return response.json();
}

/**
 * Holt eine einzelne Regel
 */
export async function getSalesBrainRule(ruleId: string): Promise<SalesBrainRule> {
  const response = await fetch(`${API_BASE}/rules/${ruleId}`, {
    credentials: "include",
  });

  if (!response.ok) {
    throw new Error(`HTTP ${response.status}`);
  }

  return response.json();
}

/**
 * Aktualisiert eine Regel
 */
export async function updateSalesBrainRule(
  ruleId: string,
  updates: Partial<{
    preferredText: string;
    note: string;
    status: "active" | "inactive";
    priority: "low" | "medium" | "high" | "critical";
  }>
): Promise<SalesBrainRule> {
  const response = await fetch(`${API_BASE}/rules/${ruleId}`, {
    method: "PATCH",
    headers: { "Content-Type": "application/json" },
    credentials: "include",
    body: JSON.stringify({
      preferred_text: updates.preferredText,
      note: updates.note,
      status: updates.status,
      priority: updates.priority,
    }),
  });

  if (!response.ok) {
    throw new Error(`HTTP ${response.status}`);
  }

  return response.json();
}

/**
 * Löscht eine Regel
 */
export async function deleteSalesBrainRule(ruleId: string): Promise<void> {
  const response = await fetch(`${API_BASE}/rules/${ruleId}`, {
    method: "DELETE",
    credentials: "include",
  });

  if (!response.ok) {
    throw new Error(`HTTP ${response.status}`);
  }
}

// =============================================================================
// MATCHING
// =============================================================================

/**
 * Findet passende Regeln für einen Kontext
 */
export async function findMatchingRules(params: {
  channel?: string;
  useCase?: string;
  leadStatus?: string;
  dealState?: string;
  inputText?: string;
  limit?: number;
}): Promise<SalesBrainRule[]> {
  const response = await fetch(`${API_BASE}/rules/match`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    credentials: "include",
    body: JSON.stringify({
      channel: params.channel,
      use_case: params.useCase,
      lead_status: params.leadStatus,
      deal_state: params.dealState,
      input_text: params.inputText,
      limit: params.limit || 5,
    }),
  });

  if (!response.ok) {
    throw new Error(`HTTP ${response.status}`);
  }

  return response.json();
}

// =============================================================================
// STATS
// =============================================================================

/**
 * Holt Sales Brain Statistiken
 */
export async function getSalesBrainStats(): Promise<{
  totalRules: number;
  userRules: number;
  teamRules: number;
  appliedThisWeek: number;
  topUseCases: { useCase: string; count: number }[];
}> {
  const response = await fetch(`${API_BASE}/stats`, {
    credentials: "include",
  });

  if (!response.ok) {
    throw new Error(`HTTP ${response.status}`);
  }

  return response.json();
}

// =============================================================================
// FEEDBACK
// =============================================================================

/**
 * Gibt Feedback zu einer angewendeten Regel
 */
export async function submitRuleFeedback(
  ruleId: string,
  feedback: {
    accepted: boolean;
    modified?: boolean;
    finalText?: string;
  }
): Promise<void> {
  const response = await fetch(`${API_BASE}/rules/${ruleId}/feedback`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    credentials: "include",
    body: JSON.stringify({
      accepted: feedback.accepted,
      modified: feedback.modified,
      final_text: feedback.finalText,
    }),
  });

  if (!response.ok) {
    throw new Error(`HTTP ${response.status}`);
  }
}

File: salesflow-app/src/backend/app/api/routes/sales_brain.py
# backend/app/api/routes/sales_brain.py
"""
╔════════════════════════════════════════════════════════════════════════════╗
║  🧠 SALES BRAIN ROUTER                                                      ║
║  API Endpoints für Teach-UI & Rule Learning                                 ║
╚════════════════════════════════════════════════════════════════════════════╝

Endpoints:
- POST /sales-brain/rules           → Neue Regel erstellen
- GET  /sales-brain/rules           → Regeln abrufen
- GET  /sales-brain/rules/{id}      → Einzelne Regel
- PATCH /sales-brain/rules/{id}     → Regel aktualisieren
- DELETE /sales-brain/rules/{id}    → Regel löschen
- POST /sales-brain/rules/match     → Passende Regeln finden
- POST /sales-brain/rules/{id}/feedback → Feedback zu Regel
- GET  /sales-brain/stats           → Statistiken
"""

from fastapi import APIRouter, HTTPException, Depends, Query
from supabase import Client
from typing import Optional

from ..schemas.sales_brain import (
    CreateRuleRequest,
    CreateRuleResponse,
    UpdateRuleRequest,
    MatchRulesRequest,
    RuleFeedbackRequest,
    SalesBrainRuleResponse,
    GetRulesResponse,
    SalesBrainStatsResponse,
)
from ...services.sales_brain import SalesBrainService, get_sales_brain_service
from ...db.deps import get_db, get_current_user, CurrentUser


# =============================================================================
# ROUTER
# =============================================================================

router = APIRouter(
    prefix="/sales-brain",
    tags=["sales-brain"],
)


# =============================================================================
# CREATE RULE
# =============================================================================

@router.post("/rules", response_model=CreateRuleResponse)
async def create_rule(
    request: CreateRuleRequest,
    current_user: CurrentUser = Depends(get_current_user),
    db: Client = Depends(get_db),
):
    """
    🧠 Erstellt eine neue Regel aus einem Override.
    
    Wird vom Teach-UI aufgerufen wenn der User "Nur für mich"
    oder "Für mein Team" auswählt.
    """
    service = get_sales_brain_service(db)
    
    result = service.create_rule(
        user_id=current_user.id,
        team_id=getattr(current_user, "team_id", None),
        scope=request.scope.value,
        override=request.override.model_dump(),
        note=request.note,
        auto_tag=request.auto_tag,
    )
    
    return CreateRuleResponse(**result)


# =============================================================================
# GET RULES
# =============================================================================

@router.get("/rules", response_model=GetRulesResponse)
async def get_rules(
    scope: Optional[str] = Query(None, pattern="^(user|team)$"),
    channel: Optional[str] = None,
    use_case: Optional[str] = None,
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    current_user: CurrentUser = Depends(get_current_user),
    db: Client = Depends(get_db),
):
    """
    Holt Regeln für den aktuellen User/Team.
    """
    service = get_sales_brain_service(db)
    
    result = service.get_rules(
        user_id=current_user.id,
        team_id=getattr(current_user, "team_id", None),
        scope=scope,
        channel=channel,
        use_case=use_case,
        page=page,
        page_size=page_size,
    )
    
    return GetRulesResponse(
        rules=[SalesBrainRuleResponse(**r) for r in result["rules"]],
        total=result["total"],
        page=result["page"],
        page_size=result["page_size"],
    )


@router.get("/rules/{rule_id}", response_model=SalesBrainRuleResponse)
async def get_rule(
    rule_id: str,
    current_user: CurrentUser = Depends(get_current_user),
    db: Client = Depends(get_db),
):
    """
    Holt eine einzelne Regel.
    """
    service = get_sales_brain_service(db)
    
    rule = service.get_rule(rule_id, current_user.id)
    
    if not rule:
        raise HTTPException(status_code=404, detail="Regel nicht gefunden")
    
    return SalesBrainRuleResponse(**rule)


# =============================================================================
# UPDATE RULE
# =============================================================================

@router.patch("/rules/{rule_id}", response_model=SalesBrainRuleResponse)
async def update_rule(
    rule_id: str,
    request: UpdateRuleRequest,
    current_user: CurrentUser = Depends(get_current_user),
    db: Client = Depends(get_db),
):
    """
    Aktualisiert eine Regel.
    """
    service = get_sales_brain_service(db)
    
    updated = service.update_rule(
        rule_id=rule_id,
        user_id=current_user.id,
        updates=request.model_dump(exclude_unset=True),
    )
    
    return SalesBrainRuleResponse(**updated)


# =============================================================================
# DELETE RULE
# =============================================================================

@router.delete("/rules/{rule_id}")
async def delete_rule(
    rule_id: str,
    current_user: CurrentUser = Depends(get_current_user),
    db: Client = Depends(get_db),
):
    """
    Löscht eine Regel.
    """
    service = get_sales_brain_service(db)
    
    service.delete_rule(rule_id, current_user.id)
    
    return {"success": True}


# =============================================================================
# MATCH RULES
# =============================================================================

@router.post("/rules/match")
async def match_rules(
    request: MatchRulesRequest,
    current_user: CurrentUser = Depends(get_current_user),
    db: Client = Depends(get_db),
):
    """
    Findet passende Regeln für einen Kontext.
    
    Wird von CHIEF verwendet um User-Regeln zu berücksichtigen.
    """
    service = get_sales_brain_service(db)
    
    rules = service.match_rules(
        user_id=current_user.id,
        team_id=getattr(current_user, "team_id", None),
        channel=request.channel,
        use_case=request.use_case,
        lead_status=request.lead_status,
        deal_state=request.deal_state,
        input_text=request.input_text,
        limit=request.limit,
    )
    
    return [SalesBrainRuleResponse(**r) for r in rules]


# =============================================================================
# FEEDBACK
# =============================================================================

@router.post("/rules/{rule_id}/feedback")
async def submit_feedback(
    rule_id: str,
    request: RuleFeedbackRequest,
    current_user: CurrentUser = Depends(get_current_user),
    db: Client = Depends(get_db),
):
    """
    Gibt Feedback zu einer angewendeten Regel.
    """
    service = get_sales_brain_service(db)
    
    service.submit_feedback(
        rule_id=rule_id,
        user_id=current_user.id,
        accepted=request.accepted,
        modified=request.modified or False,
        final_text=request.final_text,
    )
    
    return {"success": True}


# =============================================================================
# STATS
# =============================================================================

@router.get("/stats", response_model=SalesBrainStatsResponse)
async def get_stats(
    current_user: CurrentUser = Depends(get_current_user),
    db: Client = Depends(get_db),
):
    """
    Holt Sales Brain Statistiken.
    """
    service = get_sales_brain_service(db)
    
    stats = service.get_stats(
        user_id=current_user.id,
        team_id=getattr(current_user, "team_id", None),
    )
    
    return SalesBrainStatsResponse(
        total_rules=stats["total_rules"],
        user_rules=stats["user_rules"],
        team_rules=stats["team_rules"],
        applied_this_week=stats["applied_this_week"],
        top_use_cases=stats["top_use_cases"],
    )

